### PR TITLE
pseudo-p significance calculation 

### PR DESCRIPTION
**Pull request by JosiahParry**
This PR drafts a function `calculate_significance()` to provide a consistent way to calculate pseudo-p values from a reference distribution. 

It is based on the discussion at https://github.com/pysal/esda/issues/199

### CODE REVIEW

**Comment by ljwolf** [comment]
OK, done here on the logic & implementation. 

I've re-implemented the percentile-based two-sided test from scratch using `scipy.stats.scoreatpercentile`. This approach finds the percentile for the test statistic in the reference distribution and counts how many replicates are outside of (`p`, `1-p`). Over simulations, these are always 2*`directed`. Second, I modified your folding approach to fold around the mean of the replicates, rather than zero (since the expected value of local stats generally isn't zero) and kept it as a `folded` option for testing. 

I don't think we should expose the `folded` variant to the user in the classes, since the power in each direction is dependent on the symmetry of the distribution. For example, in the illustration below, the smallest replicate, when folded, is not "extreme," but this is accounted for in the percentile-based method. 

IMG_98921AE64450-1

The percentile will always equal the folded version for symmetric distributions, but the folded version becomes a directed test as skew increases. I think that the (over+under)/all is also the intended estimand of the `directed` approach, after re-reading the Hope paper referred to in #199 

If other maintainers approve these four options (`greater`, `lesser`, `two-sided`, and `directed`) for the user classes and a `folded` option for this function only (for replication/testing purposes) I can start propagating this across the user classes.

**Review by knaaptime — APPROVED** [review]
sweet. presume the stuff in __main__ gets moved to a test file or something but this is great

File: esda/significance.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        )
+    elif method == "greater":
+        p_value = (np.sum(reference_distribution <= test_stat, axis=1) + 1) / (
+            p_permutations + 1
+        )
+    elif method == "two-sided":
+        percentile = (reference_distribution < test_stat).mean(axis=1)
+        bounds = np.column_stack((1 - percentile, percentile)) * 100
+        bounds.sort(axis=1)
+        lows, highs = np.row_stack(
```
Comment (ljwolf) [inline esda/significance.py:46]: This may be able to be vectorised, but I couldn't quickly figure that out. the following did not generate the same results as below: 

```python
stats.scoreatpercentile(reference_distribution, bounds, axis=1)
```

File: esda/significance.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+from scipy import stats
+
+
+def calculate_significance(test_stat, reference_distribution, method="two-sided"):
+    """
+    Calculate a pseudo p-value from a reference distribution.
+
+    Pseudo-p values are calculated using the formula (M + 1) / (R + 1). Where R is the number of simulations and M is the number of times that the simulated value was equal to, or more extreme than the observed test statistic.
+
+    Simulated test statistics are generated through a process of conditional permutation. Conditional permutation holds fixed the value of Xi and values of neighbors are randomly sampled from X removing Xi simulating spatial randomness. This process is repeated R times to generate a reference distribution from which the pseudo-p value is calculated.
```
Comment (sjsrey) [inline esda/significance.py:11]: Perhaps a distinction might be made between global tests (unconditional permutations) and local tests (conditional permutations)?

File: esda/significance.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    reference_distribution = np.atleast_2d(reference_distribution)
+    n_samples, p_permutations = reference_distribution.shape
+    test_stat = np.atleast_2d(test_stat).reshape(n_samples, -1)
+    if method == "directed":
+        larger = (reference_distribution >= test_stat).sum(axis=1)
+        low_extreme = (p_permutations - larger) < larger
+        larger[low_extreme] = p_permutations - larger[low_extreme]
+        p_value = (larger + 1.0) / (p_permutations + 1.0)
+    elif method == "lesser":
+        p_value = (np.sum(reference_distribution >= test_stat, axis=1) + 1) / (
```
Comment (sjsrey) [inline esda/significance.py:35]: Should this be <=  ?

File: esda/significance.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            p_permutations + 1
+        )
+    elif method == "greater":
+        p_value = (np.sum(reference_distribution <= test_stat, axis=1) + 1) / (
```
Comment (sjsrey) [inline esda/significance.py:39]: Should this be >= ?

Comment (JosiahParry) [reply esda/significance.py:39]: I _think_ this order is correct because we're saying the reference distribution is less than or equal to the test stat. It may make sense to flip them just for the sake reading clarity

File: esda/significance.py (hunk 2 above)
Comment (ljwolf) [reply esda/significance.py:11]: Def, will do. This function will be used for any permutation test, local or global.

File: esda/significance.py (hunk 4 above)
Comment (ljwolf) [reply esda/significance.py:39]: Yes, I think @sjsrey is right. 

The p-value for the one sided permutation test is the fraction of replicates further out in the tail. The "alternative" refers to the alternative hypothesis that generates the test stat, not the reference distribution. 

So, for "greater", we need to count the replicates *above* the test stat... if too many are above the test stat, the p value will be large and we can't reject the null hypothesis that the test stat is *not greater* than the reference distribution.

**Comment by ljwolf** [comment]
I am still working on this, but I recall now why the implementation of an "alternative" argument was a bit trickier than I expected... because we allow for the user to ["discard" the random replicates, rather than store them](https://github.com/pysal/esda/blob/main/esda/crand.py#L329), we have to push the significance logic all the way down to the conditional randomization `numba` function, which may have significant performance implications. 

This has significant performance implications, since we're currently only counting the number of larger random stats in each innermost loop. 

It seems clear to me that 
1. if the test statistic is as large as k realizations, 
2. then there are always `2k` simulations outside of the `(k/n, (1-(k/n))` interval, plus the test statistic itself. 
3. So, the proper p-value for the two-sided test is `(2*n_greater+1)/(n_samples + 1)`, 
4. which is off from two times our current p-value by `1/n_samples`. 

If 1-4 are correct, this means we don't need to change any of the numba code. The correction can be calculated as `2*directed - (1/(n_samples+1))` __after__ the numba calculation. Do I have this right @sjsrey @knaaptime @martinfleis @jGaboardi? 

So, if we implement our current test for local stats without flipping (as greater), generate 1-p_sim (as lesser), and implement the above correction for the two-sided test (`2*p_sim - (1/(n_samples+1))`), none of the numba code needs to change. 

Is that OK w/ other maintainers?

**Comment by jGaboardi** [comment]
> This is too much stats for me to say anything useful.

Same for me.

**Review by jGaboardi — APPROVED** [review]
Since the topic is over my head, my approval is based on a general review.

**Comment by ljwolf** [comment]
One further wrinkle as well: [some global Moran tests support directed testing a binary option](https://github.com/pysal/esda/blob/main/esda/moran.py#L172) already. Notably, if `two_tailed=False`, they pick the test direction based on whether the global `I` is positive or negative. 

For us to roll-out the testing across all the classes, we need to consider if this option should be deprecated in favor of an explicit "alternative" option? Right now, there's no way to force a direction on these tests.

**Comment by sjsrey** [comment]
> I am still working on this, but I recall now why the implementation of an "alternative" argument was a bit trickier than I expected... because we allow for the user to ["discard" the random replicates, rather than store them](https://github.com/pysal/esda/blob/main/esda/crand.py#L329), we have to push the significance logic all the way down to the conditional randomization `numba` function. This may have significant performance implications, since we're currently only counting the number of larger random stats in each innermost loop.
> 
> It seems clear to me that
> 
> 1. if the test statistic is as large as `k` realizations,
> 2. then there are always `2k` simulations outside of the `(k/n, (1-(k/n))` interval, plus the test statistic itself.
> 3. So, the proper p-value for the two-sided test is `(2*n_greater+1)/(n_samples + 1)`,
> 4. which is off from two times our current p-value by `1/(n_samples+1)`.
> 
> If 1-4 are correct, this means we don't need to change any of the numba code. The correction can be calculated as `2*directed - (1/(n_samples+1))` **after** the numba calculation. Do I have this right @sjsrey @knaaptime @martinfleis @jGaboardi?
> 
> So, if we implement our current test for local stats without flipping (as greater), generate 1-p_sim (as lesser), and implement the above correction for the two-sided test (`2*p_sim - (1/(n_samples+1))`), none of the numba code needs to change.
> 
> Is that OK w/ other maintainers?

I think this is OK. 

One thing to check is if:
> 3. So, the proper p-value for the two-sided test is `(2*n_greater+1)/(n_samples + 1)`,
Should be
2(n_greater+1)/(n_samples+1)

**Comment by ljwolf** [comment]
> One thing to check is if: 

Sure, that is what I initially thought & what @JosiahParry suggested. 

The reason why I'm thinking it's actually `2*p_sim - (1/(n_permutations + 1)) is because using `2*p_sim` amounts to counting the test stat twice: `2*p_directed = 2 * (n_outside + 1)/(n+1) = (2*outside + 2)/(n+1)`. The difference will be vanishingly small as the number of permutations increases, but it's the principle... 

Thinking another way, in the percentile-based version of the test, you compute the percentile `p` for the test statistic, count how many null statistics are outside of `(p,1-p)` *and add one*, since the test stat is always at least at its own percentile. This p-value is smaller than `p_sim` by `1/(n_samples-1)`, which is what would happen in the percentile test if you counted the test statistic twice. 

the simulation code at the end of `esda/simulation.py` should illustrate?

**Comment by weikang9009** [comment]
@ljwolf I was looking at the discussions in this PR and the other related issue. The correction for the two-sided test `2*p_sim - (1/(n_samples+1))` looks correct to me.

**Comment by JosiahParry** [comment]
Thank you for the explanation @ljwolf. I think I'm almost there/onboard! 

It's worth calling out explicitly this formula can result in a p-value > 1.0 which should also be handled e.g.

``` r
p_sim = 0.65
nsim = 999

(p_corrected = (2*p_sim - (1/(nsim + 1))))
#> [1] 1.299

if (p_corrected > 1) {
  1.0
} else {
  p_corrected
}
#> [1] 1
```

Additionally, would you mind elaborating why it is `- (1/(nsim + 1))` as opposed to `+ (1/(nsim + 1))`? To me, it makes more sense to penalize smaller numbers of simulations rather than larger number of simulations. For example subtracting the second term results in smaller p values for smaller numbers of simulations and larger ones for larger numbers of simulations. 

``` r
calc_p_sim <- function(p_sim, nsim) {
  (p_corrected = (2*p_sim - (1/(nsim + 1))))

  if (p_corrected > 1) {
    1.0
  } else {
    p_corrected
  }

}

calc_p_sim(0.05, 49)
#> [1] 0.08
calc_p_sim(0.05, 99)
#> [1] 0.09
calc_p_sim(0.05, 999)
#> [1] 0.099
```

File: esda/significance.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        - 'two-sided': the observed test-statistic is an extreme value of the reference distribution.
+        - 'lesser': the observed test-statistic is small relative to the reference distribution.
+        - 'greater': the observed test-statistic is large relative to the reference distribution.
+        - 'directed': the observed test statistic is either small or large reltaive to the reference distribution.
+
+    Notes
+    -----
+
+    the directed p-value is half of the two-sided p-value, and corresponds to running the
+    lesser and greater tests, then picking the smaller significance value. This is not advised.
```
Comment (JosiahParry) [inline esda/significance.py:29]: Note that this will be untrue if the adjustment is added

File: esda/significance.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
+import numpy as np
+from scipy import stats
+
+
+def calculate_significance(test_stat, reference_distribution, method="two-sided"):
+    """
+    Calculate a pseudo p-value from a reference distribution.
+
+    Pseudo-p values are calculated using the formula (M + 1) / (R + 1). Where R is the number of simulations
+    and M is the number of times that the simulated value was equal to, or more extreme than the observed test statistic.
```
Comment (JosiahParry) [inline esda/significance.py:10]: TODO: describe the adjustment here

File: esda/significance.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    test_stat: float or numpy.ndarray
+        The observed test statistic, or a vector of observed test statistics
+    reference_distribution: numpy.ndarray
+        A numpy array containing simulated test statistics as a result of conditional permutation.
+    method: string
+        One of 'two-sided', 'lesser', or 'greater'. Indicates the alternative hypothesis.
+        - 'two-sided': the observed test-statistic is an extreme value of the reference distribution.
+        - 'lesser': the observed test-statistic is small relative to the reference distribution.
+        - 'greater': the observed test-statistic is large relative to the reference distribution.
+        - 'directed': the observed test statistic is either small or large reltaive to the reference distribution.
```
Comment (JosiahParry) [inline esda/significance.py:23]: This states that the valid arguments are `'two-sided', 'lesser', or 'greater'` however there is also a `directed` argument documented as well—though it is unclear to me how directed and two-sided are different.

**Comment by weikang9009** [comment]
> It's worth calling out explicitly this formula can result in a p-value > 1.0 which should also be handled e.g.

The formula `2*p_sim - (1/(n_samples+1))` will not result in a p-value larger than 1 since `p_sim` is equivalent to the directed case which always picks the smaller side, meaning that `p_sim` will be always no larger than 0.5. 

> Additionally, would you mind elaborating why it is - (1/(nsim + 1)) as opposed to + (1/(nsim + 1))? To me, it makes more sense to penalize smaller numbers of simulations rather than larger number of simulations. For example subtracting the second term results in smaller p values for smaller numbers of simulations and larger ones for larger numbers of simulations.

This is based on the equation for the two-sided case `p_two_sided = (2*n_greater+1)/(n_samples + 1)` and the current implementation of p_sim: `p_sim = (n_greater+1)/(n_samples + 1)` 

@ljwolf please correct me if I'm wrong

**Comment by ljwolf** [comment]
Yep, exactly right! 

@JosiahParry, think about it in terms of the percentiles, maybe that will be clearer. 

The correction term is to adjust for double counting the "observed"  stat- if you just did 2*p_sim, that computes the number of "outside" simulated statistics, plus the observed stat *twice* (as if we see it at both percentile p and 1-p) However, we actually only see the observed stat once at percentile p---the two tailed boundary in the "other" tail is derived from 1-p.

Like, imagine the number of simulations is 100. Sort them from smallest to largest. If 20 simulations are as big as the test statistic, then the test statistic is at the 80th percentile. At the (100-80)th percentile, there are n*(1-.8) *smaller* observations. 

So, we see 20 more extreme large observations, 1 test statistic, and 20 more extreme smaller observations *as if the test statistic were in the other tail*. 

2*p_sim*(n+1) would count 20 more extreme large stats, 20 more extreme small stats, and *two* test statistics-one too many. So, we need to deflate that count by 1, which deflates the p value by 1/(n+1).

**Comment by knaaptime** [comment]
for some reason this keeps reminding me of Downey's [Euro Problem](https://allendowney.github.io/ThinkBayes2/chap04.html#the-euro-problem) example, probably just because it reminds me "the statistician"'s default test is always two-sided
Screenshot 2024-05-02 at 1 14 35 PM

**Comment by JosiahParry** [comment]
Going through my stale PRs. Has this been implemented and documented in the most recent releases?  I think some long-ish form documentation on the "improved" calculation of pseudo-p values would be quite great.

**Comment by ljwolf** [comment]
Hi! back again :) this has not been forgotten---it is the highest priority for me when I have development time. 

I needed to move to a pure numpy version of the two-sided percentile test in order to push it down into the `numba.njit()` inner loop of `esda.crand()`. I will push that code up shortly, bandwidth permitting. My intention is then to write an `numba.njit()` compatible `esda.significance._permutation_test()` function, and send both `esda.crand.parallel_crand()` and `esda.significance.permutation_test()` to that when calculating the permutation test. 

We'll need to do a little profiling, but hopefully there is not *much* gain from doing it inline vs. calling another jitted function on singly-typed input.

File: esda/significance.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+        """
+        decorator mimicking numba.jit
+        """
+
+        def intercepted_function(f, *f_args, **f_kwargs):
+            return f
+
+        return intercepted_function
+
+    njit = jit
```
Comment (martinfleis) [inline esda/significance.py:18]: We've been typically doing this as below to avoid reimplementation of decorator in every package but it does not matter much.

```py
try:
    from numba import njit
except (ModuleNotFoundError, ImportError):
    from libpysal.common import jit as njit
```

File: esda/significance.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        reference_distribution = np.abs(reference_distribution - means)
+        p_value = ((reference_distribution >= test_stat).sum(axis=1) + 1) / (
+            p_permutations + 1
+        )
+    else:
+        p_value = np.ones((n_samples, 1))*np.nan
+    return p_value
+
+
+if __name__ == "__main__":
```
Comment (martinfleis) [inline esda/significance.py:111]: All below should be turned into tests but I am sure you're aware. Leaving a note to avoid this being merged like this.

Comment (ljwolf) [reply esda/significance.py]: Yes, I've specified what tests ought to look like now off of this. Will be done w/ benchmarking. We need to prepare both correctness and speed tests for the permutation stuff, since it's so finicky and easy to break.

**Comment by ljwolf** [comment]
@weikang9009 notes correctly that we will also need to update the notebooks where local/global statistics are used before merging this.

**Comment by ljwolf** [comment]
the affected notebook has been fixed and #376 has been addressed!
